### PR TITLE
feat(jwt) enhance JWT error notifications with details

### PR DESCRIPTION
**Pull request by mihhu**

### CODE REVIEW

File: lang/main-enGB.json [hunk 1]
```diff
@@ -280,7 +280,23 @@
         "streamKey": "Live stream key",
         "thankYou": "Thank you for using {{appName}}!",
         "token": "token",
-        "tokenAuthFailed": "Sorry, you're not allowed to join this call.",
+        "tokenAuthFailed": "Sorry, you're not allowed to join this call. Possible reasons: {{reason}}",
+        "tokenAuthFailedReason": {
```
Comment (saghul) [inline lang/main-enGB.json]: You don't need to add these, if a translation is missing then it falls back to main.json.

Unrelated: In fact, en-GB is probably dumb to have at this point, since it's the same as the default...

File: lang/main.json [hunk 2]
```diff
@@ -439,7 +439,23 @@
         "streamKey": "Live stream key",
         "thankYou": "Thank you for using {{appName}}!",
         "token": "token",
-        "tokenAuthFailed": "Sorry, you're not allowed to join this call.",
+        "tokenAuthFailed": "Sorry, you're not allowed to join this call. Possible reasons: {{reason}}",
+        "tokenAuthFailedReason": {
+            "audInvalid": "Ivalid `aud` value. It should be `jitsi`.",
+            "contextNotFound": "The `context` object is missing from the payload.",
+            "expInvalid": "Invalid `exp` value.",
+            "featureInvalid": "Invalid feature",
```
Comment (saghul) [inline lang/main.json:447]: Can we say something more here?

File: lang/main.json [hunk 3]
```diff
@@ -439,7 +439,23 @@
...
+            "featuresNotFound": "The `features` object is missing from the payload.",
+            "featureValueInvalid": "Invalid value for feature",
+            "headerOrPayload": "Missing the header or payload.",
+            "issInvalid": "Invalid `iss` value. It should be `chat`.",
```
Comment (saghul) [inline lang/main.json:451]: Same as above, "chat" is a JaaS convention. Maybe we can tell users to check "the documentation" so it's generic enough?

File: react/features/base/jwt/functions.ts [hunk 4]
```diff
@@ -1,3 +1,6 @@
+// eslint-disable-next-line import/order
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:1]: Why is this needed? It's properly sorted...

File: react/features/base/jwt/functions.ts [hunk 5]
```diff
@@ -75,26 +78,24 @@ function isValidUnixTimestamp(timestamp: number | string) {
...
 
     try {
         const header = jwtDecode(jwt, { header: true });
         const payload = jwtDecode(jwt);
 
         if (!header || !payload) {
-            errors.push('- Missing header or payload');
+            errors = {
+                ...errors,
+                '- Missing header or payload': i18n.t('dialog.tokenAuthFailedReason.headerOrPayload')
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:97]: Not a fan of this. We could instead return an array or constants, and then render the translated messages in the notification, can't we?

I imagine something like:

```
const JWT_VALIDATION_ERRORS: {
    MISSING_HEADER: 'missing-header',
    MISSING_PAYLOAD: 'missing-payload',
    MISSING_KID: 'missing-kid',
    ...
};
```

Generally speaking we try not to pass translated strings around, but to translate them right where they are rendered.

File: react/features/base/jwt/functions.ts [hunk 6]
```diff
@@ -165,15 +197,28 @@ export function validateJwt(jwt: string) {
                         && featureValue !== 'true'
                         && featureValue !== 'false'
                     ) {
-                        errors.push(`- Invalid value for feature: ${feature}`);
+                        errors = {
+                            ...errors,
+                            [`- Invalid value for feature: ${feature}`]:
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:202]: This could be expressed as a `invalid-value-for-feature-THEFEATURE` constant, following my suggestion above.

File: react/features/base/jwt/functions.ts [hunk 7]
```diff
@@ -165,15 +197,28 @@ export function validateJwt(jwt: string) {
...
+                                `${i18n.t('dialog.tokenAuthFailedReason.featureValueInvalid')}: ${feature}.`
+                        };
                     }
                 } else {
-                    errors.push(`- Invalid feature: ${feature}`);
+                    errors = {
+                        ...errors,
+                        [`- Invalid feature: ${feature}`]:
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:209]: Ditto.

File: react/features/base/jwt/functions.ts [hunk 8]
```diff
@@ -165,15 +197,28 @@ export function validateJwt(jwt: string) {
...
+                            `${i18n.t('dialog.tokenAuthFailedReason.featureInvalid')}: ${feature}.`
+                    };
                 }
             });
         }
     } catch (e: any) {
-        errors.push(e ? e.message : '- unspecified jwt error');
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:176]: I'd add a logger.error here.

File: react/features/base/jwt/functions.ts [hunk 9]
```diff
@@ -165,15 +197,28 @@ export function validateJwt(jwt: string) {
...
+        const message = e?.message;
+
+        errors = {
+            ...errors,
+            [message || '- unspecified jwt error']: message || i18n.t('dialog.tokenAuthFailedReason.unspecifiedError')
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:220]: I wouldn't pass the error message here. The catch-all is there mostly as a safely net. If it trips it's probably a bug in our parsing code I think.

**Review by saghul — COMMENTED** [review]
Did a first pass, PTAL!

File: react/features/base/jwt/functions.ts (hunk 4 above)
Comment (mihhu) [reply react/features/base/jwt/functions.ts:1]: Because we've got a conflict with the comment below, it requires a space that triggers this error.

File: react/features/base/jwt/functions.ts (hunk 5 above)
Comment (mihhu) [reply react/features/base/jwt/functions.ts:97]: > to translate them right where they are rendered.
Yes, that was my initial approach, with the enum like object. But then I needed arguments to my description arguments and I started twisting the Notification.tsx implementation to translate arguments and stumbled into the scenario in which we might translate something we don't want translated (your occasional odd username). I figured this is the safest approach.

I chose to keep the keys from the initial implementation to display them directly in the logs, because I found it odd to translate that part. WDYT?

File: react/features/base/jwt/functions.ts (hunk 6 above)
Comment (mihhu) [reply react/features/base/jwt/functions.ts:202]: Are you sure? There are 15 of those atm..

File: react/features/base/conference/middleware.any.ts [hunk 10]
```diff
@@ -415,18 +410,23 @@ function _logJwtErrors(message: string, state: IReduxState) {
  * @returns {Object} The value returned by {@code next(action)}.
  */
 function _connectionFailed({ dispatch, getState }: IStore, next: Function, action: AnyAction) {
-    _logJwtErrors(action.error.message, getState());
-
     const { connection, error } = action;
+    const { jwt } = getState()['features/base/jwt'];
+    let errorKeys: Object = {};
```
[the remark this answers is not in the document]
Comment (mihhu) [reply react/features/base/conference/middleware.any.ts:415]: Yes, thanks, leftovers from refactoring

File: lang/main.json (hunk 2 above)
Comment (mihhu) [reply lang/main.json:447]: Like "feature doesn't exist"?

File: react/features/base/jwt/functions.ts (hunk 6 above)
Comment (saghul) [reply react/features/base/jwt/functions.ts:203]: Why not? It's a dynamic field, we wouldn't have a dedicated constant for each bur taher a prefix, so if it starts with `invalid-value-for-feature-` we know what comes next.

File: react/features/base/jwt/functions.ts (hunk 5 above)
Comment (saghul) [reply react/features/base/jwt/functions.ts:98]: Can't the translation happen at the time of dispatch then?

File: lang/main.json (hunk 3 above)
Comment (saghul) [reply lang/main.json:451]: Sorry I hadn't notied this is inside an if (jaas) check so it's ok.

File: lang/main.json (hunk 2 above)
Comment (saghul) [reply lang/main.json:447]: Yeah, sounds good.

File: lang/main.json [hunk 11]
```diff
@@ -439,7 +439,23 @@
         "streamKey": "Live stream key",
         "thankYou": "Thank you for using {{appName}}!",
         "token": "token",
-        "tokenAuthFailed": "Sorry, you're not allowed to join this call.",
+        "tokenAuthFailed": "Sorry, you're not allowed to join this call. Possible reasons: {{reason}}",
+        "tokenAuthFailedReason": {
+            "audInvalid": "Ivalid `aud` value. It should be `jitsi`.",
```
[the remark this answers is not in the document]
Comment (saghul) [reply lang/main.json:444]: Disregard, this check is inside the if (jaas) part so it's OK.

File: react/features/base/jwt/functions.ts (hunk 5 above)
Comment (mihhu) [reply react/features/base/jwt/functions.ts:98]: You mean in the middleware?

Comment (mihhu) [reply react/features/base/jwt/functions.ts:98]: And have the localized keys as the value in the `JWT_VALIDATION_ERRORS` object?

Comment (saghul) [reply react/features/base/jwt/functions.ts:98]: Yes in the middleware, where we dispatch the showNotification action.

> And have the localized keys as the value in the `JWT_VALIDATION_ERRORS` object?

I don't think you can do this because the i18n backend needs to be initialized and it won't be at import time so you can't define a constant.

Comment (mihhu) [reply react/features/base/jwt/functions.ts:98]: I meant to have what you described above with the keys instead:
```
const JWT_VALIDATION_ERRORS: {
    MISSING_HEADER: 'headerNotFound',
    MISSING_PAYLOAD: 'headerNotFound',
    MISSING_KID: 'kidNotFound',
    ...
};
```

And compose the locale key in the dispatch as you said, by iterating over the errors and calling something like t(\`dialog.tokenAuthFailedReason.${error[idx]}\`)

Comment (saghul) [reply react/features/base/jwt/functions.ts:98]: Yeah, makes sense!

File: lang/main-enGB.json (hunk 1 above)
Comment (mihhu) [reply lang/main-enGB.json]: I've updated the existing key, tokenAuthFailed, because otherwise it would've omitted the whole thing. Should I do this for all languages?

**Comment by mihhu** [comment]
Please Jenkins test this please

**Comment by mihhu** [comment]
Please Jenkins test this please

File: react/features/base/jwt/functions.ts [hunk 12]
```diff
@@ -165,15 +167,21 @@ export function validateJwt(jwt: string) {
                         && featureValue !== 'true'
                         && featureValue !== 'false'
                     ) {
-                        errors.push(`- Invalid value for feature: ${feature}`);
+                        errors.push({
+                            key: JWT_VALIDATION_ERRORS.FEATURE_VALUE_INVALID,
+                            args: { feature }
```
Comment (saghul) [inline react/features/base/jwt/functions.ts:172]: Great stuff, better than my initial suggestion for sure. ❤️

**Review by saghul — APPROVED** [review]
Awesome job! 🚀
